
% CIBM poster, based on design by Nathaniel Johnston, August 2009
% http://www.nathanieljohnston.com/2009/08/latex-poster-template/

\documentclass[final]{beamer}

\usepackage[scale=1.24]{beamerposter}
\usepackage{graphicx}			% allows us to import images
\usepackage{tikz}
\usepackage{url}

%-----------------------------------------------------------
% Custom commands that I use frequently
%-----------------------------------------------------------

\newcommand{\bb}[1]{\mathbb{#1}}
\newcommand{\cl}[1]{\mathcal{#1}}
\newcommand{\fA}{\mathfrak{A}}
\newcommand{\fB}{\mathfrak{B}}
\newcommand{\Tr}{{\rm Tr}}
\newtheorem{thm}{Theorem}

%-----------------------------------------------------------
% Define the column width and poster size
% To set effective sepwid, onecolwid and twocolwid values, first choose how many columns you want and how much separation you want between columns
% The separation I chose is 0.024 and I want 4 columns
% Then set onecolwid to be (1-(4+1)*0.024)/4 = 0.22
% Set twocolwid to be 2*onecolwid + sepwid = 0.464
%-----------------------------------------------------------



\definecolor{Sinput}{rgb}{0.75,0.19,0.19}
\definecolor{Soutput}{rgb}{0,0,0}
\definecolor{Scode}{rgb}{0.75,0.19,0.19}


\setbeamertemplate{sections/subsections in toc}[square] % This is to remove the symbols of the slides
\setbeamertemplate{navigation symbols}{} % This is to convert the "3d" spheres to squares on toc
\setbeamertemplate{itemize item}[square] % This is to convert the "3d" spheres to squares on enumerate





\newlength{\sepwid}
\newlength{\onecolwid}
\newlength{\twocolwid}
\setlength{\paperwidth}{72in}
\setlength{\paperheight}{48in}
\setlength{\sepwid}{0.024\paperwidth}
\setlength{\onecolwid}{0.22\paperwidth}
\setlength{\twocolwid}{0.464\paperwidth}
\setlength{\topmargin}{-0.5in}
\usetheme{confposter}
\usepackage{exscale}




%-----------------------------------------------------------
% The next part fixes a problem with figure numbering. Thanks Nishan!
% When including a figure in your poster, be sure that the commands are typed in the following order:
% \begin{figure}
% \includegraphics[...]{...}
% \caption{...}
% \end{figure}
% That is, put the \caption after the \includegraphics
%-----------------------------------------------------------

\usecaptiontemplate{
\small
\structure{\insertcaptionname~\insertcaptionnumber:}
\insertcaption}

%-----------------------------------------------------------
% Define colours (see beamerthemeconfposter.sty to change these colour definitions)
%-----------------------------------------------------------

\setbeamercolor{block title}{fg=NavyBlue,bg=white}
\setbeamercolor{block body}{fg=black,bg=white}
\setbeamercolor{block alerted title}{fg=white,bg=dblue!70}
\setbeamercolor{block alerted body}{fg=black,bg=dblue!10}

%-----------------------------------------------------------
% Name and authors of poster/paper/research
%-----------------------------------------------------------


\title{Segvis: A package for visualization of high throughout sequencing data along genomic segments}
\author{Rene Welch and S\"{u}nd\"{u}z Kele\c{s}}
\institute{Department of Statistics, University of Wisconsin - Madison
\hspace{1 cm} 
  Department of Biostatistics and Medical Informatics, University of Wisconsin - Madison}
\date{October 2014}


\usepackage{tikz}

\tikzstyle{sqq} = [rectangle, draw ,fill = MidnightBlue!25,
    text width=6.7em, text centered, rounded corners, minimum height=10em,line width = 1pt,draw = black!60!blue!95]
\tikzstyle{sqq2} = [rectangle, draw ,fill = SeaGreen!25,
    text width=6.7em, text centered, rounded corners, minimum height=10em,line width = 1pt,draw = black!60!green!95]



\tikzstyle{line} = [draw,line width = 2pt, -latex',fill =black!60!blue!95,draw = black!60!blue!95]
\tikzstyle{fancytitle} =[fill=black!60!blue!95, text=white,rounded corners,text width = 55 pt,text centered]

\usetikzlibrary{shapes,arrows}
\usetikzlibrary{mindmap,trees}

\usepackage{listings}  

\definecolor{mygreen}{rgb}{0,0.6,0}
\definecolor{mygray}{rgb}{0.5,0.5,0.5}
\definecolor{mymauve}{rgb}{0.58,0,0.82}


\lstset{ %
  backgroundcolor=\color{white},   % choose the background color; you must add \usepackage{color} or \usepackage{xcolor}
  basicstyle=\normalsize,        % the size of the fonts that are used for the code
  breakatwhitespace=false,         % sets if automatic breaks should only happen at whitespace
  breaklines=true,                 % sets automatic line breaking
  captionpos=b,                    % sets the caption-position to bottom
  commentstyle=\color{mygreen},    % comment style
  deletekeywords={...},            % if you want to delete keywords from the given language
  escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  extendedchars=true,              % lets you use non-ASCII characters; for 8-bits encodings only, does not work with UTF-8
  frame=single,                    % adds a frame around the code
  keepspaces=true,                 % keeps spaces in text, useful for keeping indentation of code (possibly needs columns=flexible)
  keywordstyle=\color{NavyBlue},       % keyword style
  language=Octave,                 % the language of the code
  morekeywords={*,...},            % if you want to add more keywords to the set
  numbers=left,                    % where to put the line-numbers; possible values are (none, left, right)
  numbersep=5pt,                   % how far the line-numbers are from the code
  numberstyle=\tiny\color{mygray}, % the style that is used for the line-numbers
  rulecolor=\color{black},         % if not set, the frame-color may be changed on line-breaks within not-black text (e.g. comments (green here))
  showspaces=false,                % show spaces everywhere adding particular underscores; it overrides 'showstringspaces'
  showstringspaces=false,          % underline spaces within strings only
  showtabs=false,                  % show tabs within strings adding particular underscores
  stepnumber=2,                    % the step between two line-numbers. If it's 1, each line will be numbered
  stringstyle=\color{mymauve},     % string literal style
  tabsize=2,                       % sets default tabsize to 2 spaces
  title=\lstname                   % show the filename of files included with \lstinputlisting; also try caption instead of title
}

%-----------------------------------------------------------
% Start the poster itself
%-----------------------------------------------------------
% The \rmfamily command is used frequently throughout the poster to force a serif font to be used for the body text
% Serif font is better for small text, sans-serif font is better for headers (for readability reasons)
%-----------------------------------------------------------


\begin{document}
\begin{frame}[fragile,t]
  \begin{columns}[t] % the [t] option aligns the column's content at the top

    \begin{column}{\sepwid}\end{column}			% empty spacer column
    \begin{column}{\onecolwid}


\begin{alertblock}{What is Segvis?}
  
  {\color{NavyBlue} \textbf{Segvis}} is a tool for visualization of
  high throughout sequencing raw data. The minimum input to the
  package includes:

\begin{enumerate}
\item One or more bam files of aligned read data (e.g. from ChIP-seq
  experiments)

\item Coordinates for regions of interest

\end{enumerate}

{\color{NavyBlue}\textbf{Segvis}} provides different tools to
summarize and visualize these data, including but not limited to the
following tasks:

\begin{itemize}
\item Extract read data of specified input regions
\item Plot data from different files (conditions) accross the same set
  of regions, e.g. peak plots for ChIP-seq
\item Plot central measures for a set of regions with the same width
\item Subset this regions according to user defined annotations
%\item Plot the heatmap of signal curves accross regions separated by
 % annotation
\end{itemize}

\end{alertblock}
\vskip2ex
\begin{block}{Basic input}

\begin{enumerate}


\item A set of regions

\begin{table}[H]
\begin{tabular}{c | c|  c| c | c}
chr1& 114889057 & 114889538 & ... & 240 \\ 
chr1& 225662556 & 225663044 & ... & 249 \\
chr1& 150951878 & 150952345 & ... & 259 \\
chr1& 17036198  & 17036690  & ... & 255 \\
chr1& 35318207  & 35318710  & ... & 262 \\
chr1& 204776085 & 204776784 & ... & 376 \\
chr1& 33177715  & 33178175  & ... & 239 \\
chr1& 19239498  & 19239983  & ... & 230 \\
chr1& 38455665  & 38456081  & ... & 185 \\
chr1& 154989953 & 154990397 & ... & 221 
\end{tabular}
\caption{First 10 peaks in ENCODE generated narrowPeak file (content table)}
\label{tab1}
\end{table}

\item And a set of at least 1 {\bf bam} file

\end{enumerate}

\end{block}
      \vskip2ex
\begin{block}{Creating input regions}
        
\begin{lstlisting}[language =R]
  content = read.table(file = "file.narrowPeak",
stringsAsFactors=FALSE)
  peaks_ctcf = GRanges(seqnames = content[,1],
    ranges =IRanges(start=content[,2],
    end =content[,3]),strand   = '*')
  peaks_ctcf$summit = start(peaks_ctcf) + content[,10]
\end{lstlisting}%$

\end{block}

      \vskip2ex
\begin{block}{The {\color{orange} profile} and {\color{orange}profileMatrix} objects}
      
  \begin{lstlisting}[language=R]
    
  name = "H3k27ac"; file ="reads_H3k27ac_first3chr.sort.bam"
  maxBw = 501;  windowExt = 1000 ;  multi.core = 8
  fl = 200; # only for SET
  ourProfile = Profile(regionName = name,file = file,fileFormat= "bam",maxBandwidth = maxBw,fragLen = fl,remChr = "")

  start(peaks_ctcf) = peaks_ctcf$summit - windowExt
  end(peaks_ctcf) = peaks_ctcf$summit + windowExt
  regions(ourProfile) = peaks_ctcf

  ourProfile = loadReads(ourProfile,multi.core)
  ourProfile = matchReads(ourProfile,multi.core)
  ourProfile = getCoverage(ourProfile,multi.core)

  \end{lstlisting}
\end{block}

      \vskip2ex
\end{column}

%%% col #1

\begin{column}{\sepwid}\end{column}			
% empty spacer column
\begin{column}{\twocolwid}					
% create a two-column-wide column and then we will split it up later
\begin{columns}[t,totalwidth=\twocolwid]	% split up that two-column-wide column
\begin{column}{\onecolwid}\vspace{-.69in}
\begin{block}{The {\color{orange} profileMatrixList} object}

\vskip2ex

We repeat the procedure in the {\color{NavyBlue}\bf  profile} box for all read files, and create the {\color{NavyBlue}\bf  profileMatrices} by:
\vskip2ex
  \begin{lstlisting}[language = R]
  bandwidth = 1
  ourMatrices = 
lapply(ourProfiles, ProfileMatrix,bandwidth,multi.core)   
  \end{lstlisting}

  We normalize them and create a {\color{NavyBlue} \bf profileMatrixList}:

\vskip2ex
  \begin{lstlisting}[language = R]
  ourMatrices = lapply(ourMatrices,normalize.matrix)
  ourList = ProfileMatrixList(ourMatrices)
  \end{lstlisting}

We plot a single peak by using:

  \vskip2ex
  \begin{lstlisting}[language = R]
i=10
st =starts[i]
plot.profiles(ourList,condition = seqnames == "chr1" & start == st,coord = seq(starts[i],ends[i]))

  \end{lstlisting}

\end{block}
\end{column}
\begin{column}{\onecolwid}\vspace{-.69in}
\begin{block}{Figure 1}
\vskip3ex
   \begin{figure}[H]
    \centering
    \includegraphics[width = .45\textwidth,page = 1]{figs/fig0.pdf}
    \includegraphics[width = .45\textwidth,page = 2]{figs/fig0.pdf}
  \end{figure}
  \begin{figure}[H]
  \centering
    \includegraphics[width = .45\textwidth,page = 3]{figs/fig0.pdf}
    \includegraphics[width = .45\textwidth,page = 4]{figs/fig0.pdf}
  \end{figure}

\end{block}
\end{column}
\end{columns}

\vskip2.5ex
\begin{alertblock}{Some thoughts on {\bf plot.profiles}}

\begin{itemize}

\item The output of {\bf \color{NavyBlue} plot.profiles} is a {\bf \color{NavyBlue} ggplot} object. Thus is very easy to interact with it

\item It has 3 arguments besides the {\color{NavyBlue}\bf  profileMatrixList}: {\color{NavyBlue} coord} (the genomic coordinates),{\color{NavyBlue} trim} (a value to calculate the trimmed mean for a collectio of coverage) and {\color{NavyBlue} condition} (to subset the whole list of regions)

\end{itemize}

      \end{alertblock}
      \begin{columns}[t,totalwidth=\twocolwid]
        \begin{column}{\onecolwid}
\begin{block}{Adding annotations}
\vskip2ex

The regions of interest are an element of the {\color{NavyBlue} \bf profileMatrix}. Thus, we can add an annotation as part of the regions by:

\vskip2ex
\begin{lstlisting}[language =R]
content = read.table(file = "dnase.narrowPeak"),stringsAsFactors=FALSE)
dnase_regions = GRanges(seqnames = content[,1],ranges =IRanges(start =content[,2],end =content[,3]),strand = '*')

content = read.table(file = "Pol2b.narrowPeak"),stringsAsFactors=FALSE)
pol2b = GRanges(seqnames = content[,1],ranges =IRanges(start =content[,2],end =content[,3]),strand = '*')

regions(ourMatrix)$dnase = countOverlaps(regions(ourMatrix),dnase_regions)
regions(ourMatrix)$pol2b = countOverlaps(regions(ourMatrix),pol2b)
\end{lstlisting}

And subset by:
\vskip2ex

\begin{lstlisting}[language =R]
  plot.profiles(ourList,coord = -windowExt:windowExt) # no subset
  plot.profiles(ourList,coord = -windowExt:windowExt,condition = dnase > 0 & pol2b > 0)
\end{lstlisting}

\end{block}
\end{column}
\begin{column}{\onecolwid}
\begin{block}{Figure 2}
 \vskip3ex
  \begin{figure}[H]
    \centering
    \includegraphics[width = .45\textwidth,page = 2]{figs/fig3.pdf}
    \includegraphics[width = .45\textwidth,page = 3]{figs/fig3.pdf}
  \end{figure}
  \begin{figure}[H]
  \centering
    \includegraphics[width = .45\textwidth,page = 4]{figs/fig3.pdf}
    \includegraphics[width = .45\textwidth,page = 5]{figs/fig3.pdf}
  \end{figure}

\end{block}
\end{column}
\end{columns}
\end{column}

  \begin{column}{\sepwid}\end{column} % empty spacer column
  \begin{column}{\onecolwid}
    \begin{block}{Formatting the plots}

We can modify the format of the plots by adding element to them:

\vskip2ex
\begin{lstlisting}
# Base plot
q2 = plot.profiles(ourList,coord= -windowExt:windowExt)

# Adding a dashed line in the middle
q2 + geom_vline(xintercept = 0,linetype = "dashed")
  
# Changing colors
q2 + scale_color_brewer(palette = "Dark2")

# Black and white theme
q2 + theme_bw()

\end{lstlisting}

    \end{block}
    \vskip2ex

    \begin{block}{Figure 3}
       \vskip3ex
  \begin{figure}[H]
    \centering
    \includegraphics[width = .45\textwidth,page = 1]{figs/fig2.pdf}
    \includegraphics[width = .45\textwidth,page = 2]{figs/fig2.pdf}
  \end{figure}
  \begin{figure}[H]
  \centering
    \includegraphics[width = .45\textwidth,page = 3]{figs/fig2.pdf}
    \includegraphics[width = .45\textwidth,page = 4]{figs/fig2.pdf}
  \end{figure}


    \end{block}

    \begin{block}{For Further Information}
The package is in {\color{RoyalBlue} \url{https://github.com/welch16/profile}}
    \end{block}
    \vskip2ex
    \begin{block}{Acknowledgements}

Acknowledgement of Support: R.W. is supported by CONACYT (Mexico's National Council for Science and Technology) by the graduate studies grant 215196



    \end{block}
  \end{column}
  \begin{column}{\sepwid}\end{column}			% empty spacer column
 \end{columns}
\end{frame}
\end{document}
